### ISSUE TITLE
inputs = numpy.array(inputs_list, ndmin=2).T

### ISSUE THREAD

**Issue by Gerrit02** [makeyourownneuralnetwork/makeyourownneuralnetwork#27]
Hello,
I do not understad the part:
inputs = numpy.array(inputs_list, ndmin=2).T

Will it automatically converted in a two dimensional 28x28 array.
But in the inputs_list you still have also the label no?

**Comment by makeyourownneuralnetwork** [comment]
hi @Gerrit02

That line converts from a python list to a numpy array. By default it will have dimensions (784,) with the second dimension undefined. By setting ndmin=2, it forces the dimensions to be (784,1) which avoids any problems later for example when using multiplication.

The data is never converted to a 28x28 array for the purposes of training or querying - there's no need. It is only done when the image is displayed in the other code examples.

Does that help?

**Comment by Gerrit02** [comment]
Hoy Rashid,
not 100% clear.
I expect to have then (785,) with the second dimension undefined and when forces the dimensions then I expect to have (785,1). 
Because as said in the inputs_list you have also the label. So total of 785 digits and not 784 ?
Am I wrong somewhere?

**Comment by Gerrit02** [comment]
By the way, a very great book !!!

**Comment by makeyourownneuralnetwork** [comment]
hi @Gerrit02 - good question.

You can do this many ways, but here we have already separated out the labels from the data. Here the train() function is defined as 

` train(self, inputs_list, targets_list):
`

So the `inputs_list` is the 784 values, and the `targets_list` is the 10 values with one being switched on.

You can see elsewhere in the code when the data is read from the files:

```
        # create the target output values (all 0.01, except the desired label which is 0.99)
        targets = numpy.zeros(output_nodes) + 0.01
        # all_values[0] is the target label for this record
        targets[int(all_values[0])] = 0.99
```

Here is where the **targets** list is created. So if the label is 2, targets will be [0, 0, 1, 0, 0, 0, 0, 0, 0, 0]. In this case we use 0.01 and 0.99 to help with the training. (sigmoids can't reach 1)

**Comment by makeyourownneuralnetwork** [comment]
if you like the book - please do leave an amazon review! it helps me invest time in my next books.

**Comment by Gerrit02** [comment]
Hello,
I will certainly left a comment on amazon. I will let you know when done.
However one more question for now because still isn't something very clear for me.

Where or how do you define or say that inputs_list = 784.
When I take a look at the code the only one I see the inputs_list is in the line:
inputs = numpy.array(inputs_list, ndmin=2).T

But I can't find or see anything in the code the inputs_list is 784 :-(

**Comment by makeyourownneuralnetwork** [comment]
Each row of the training data has 785 numbers. That is the label, and 784 pixel values.

The following code shows how the label is split from the rest:
```
# split the record by the ',' commas
all_values = record.split(',')
# scale and shift the inputs
inputs = (numpy.asfarray(all_values[1:]) / 255.0 * 0.99) + 0.01
```

Because the data file is a CSV file, the numbers are split by the comma (comma separated values). This is assigned to a **list** called `all_values`.

The inputs are everything except the first item. That's why the above says `all_values[1:]` which is everything excluding the zero-th item.

The label itsel is just the first zero-th item:

```
# all_values[0] is the target label for this record
targets[int(all_values[0])] = 0.99
```

You can see `all_values[0]`.

Does that help?

**Comment by makeyourownneuralnetwork** [comment]
also have a look at this [blog post](http://makeyourownneuralnetwork.blogspot.co.uk/2015/03/the-mnist-dataset-of-handwitten-digits.html) which explains the dataset in more detail and why the training and test data has a labal and then 28x28=784 values.

**Comment by Gerrit02** [comment]
Hoy Rashid,
I understand all of the above you explain but I'm affraid it is not the answer I need :-(.
I will recall.

Following part of the code:
# query the neural network
    def query(self, inputs_list):
        # convert inputs list to 2d array
        inputs = numpy.array(inputs_list, ndmin=2).T

If I understand wel inputs gets the value of a 2d array namely [784,1]
But my question: where or how is decided that inputs_list is 784
In other words, when gets the inputs_list the value 784? Where or how does this happen in the code?

Tnx,

**Comment by makeyourownneuralnetwork** [comment]
I see your question now.

The following code creates a new object of the neural network class:

```
# number of input, hidden and output nodes
input_nodes = 784
..
..
# create instance of neural network
n = neuralNetwork(input_nodes,hidden_nodes,output_nodes, learning_rate)
```

You can see the parameter `input_nodes = 784` used to create the neural network, `n`.

Inside the neural network setup code we use this paramter:

```
# initialise the neural network
    def __init__(self, inputnodes, hiddennodes, outputnodes, learningrate):
        # set number of nodes in each input, hidden, output layer
        self.inodes = inputnodes
        self.hnodes = hiddennodes
        self.onodes = outputnodes
```

You can see the number of input nodes being set `self.inodes = inputnodes`, which was passed at initialisation time as input_nodes. The code which actually creates the weights matrix uses this number to set the correct size:

```
self.wih = numpy.random.normal(0.0, pow(self.inodes, -0.5), (self.hnodes, self.inodes))
```

Why 784? Because that's number of data values each record of the training and test sets has. If we had a different data set, we would have to decide the number of input nodes, and create a neural network with that parameter - eg:

```
# create instance of neural network
n = neuralNetwork(10, 10, 5, 0.1)
```

creates a neural network with 10 inout nodes, 10 middle-layer nodes, and 5 output layer nodes, with a learnign rate of 0.1

Does this help?

**Comment by Gerrit02** [comment]
Yes tnx, this makes more sence to me

**Comment by Gerrit02** [comment]
By the way, I left a comment on Bol.com
It's the site where I ordered this book and where I have an account.

**Comment by makeyourownneuralnetwork** [comment]
Thanks!
